server/websocket.go

```diff
@@ -465,83 +465,23 @@ func (p *Plugin) handleLeave(us *session, userID, connID, channelID string) erro
 	return nil
 }
 
-func (p *Plugin) handleJoin(userID, connID, channelID, title, threadID string) error {
-	p.LogDebug("handleJoin", "userID", userID, "connID", connID, "channelID", channelID)
+func (p *Plugin) handleJoin(userID, connID string, req CallStartRequest) error {
+	p.LogDebug("handleJoin", "userID", userID, "connID", connID, "channelID", req.ChannelID)
 
-	// We should go through only if the user has permissions to the requested channel
-	// or if the user is the Calls bot.
-	if !(p.isBot(userID) || p.API.HasPermissionToChannel(userID, channelID, model.PermissionCreatePost)) {
-		return fmt.Errorf("forbidden")
-	}
-	channel, appErr := p.API.GetChannel(channelID)
-	if appErr != nil {
-		return appErr
-	}
-	if channel.DeleteAt > 0 {
-		return fmt.Errorf("cannot join call in archived channel")
-	}
-
-	if threadID != "" {
-		post, appErr := p.API.GetPost(threadID)
-		if appErr != nil {
-			return appErr
-		}
-
-		if post.ChannelId != channelID {
-			return fmt.Errorf("forbidden")
-		}
-
-		if post.DeleteAt > 0 {
-			return fmt.Errorf("cannot attach call to deleted thread")
-		}
-
-		if post.RootId != "" {
-			return fmt.Errorf("thread is not a root post")
-		}
+	channel, err := p.getChannelForCall(userID, req)
+	if err != nil {
+		return fmt.Errorf("failed to get channel for call: %w", err)
 	}
 
 	state, prevState, err := p.addUserSession(userID, connID, channel)
 	if err != nil {
 		return fmt.Errorf("failed to add user session: %w", err)
 	} else if state.Call == nil {
 		return fmt.Errorf("state.Call should not be nil")
-	} else if len(state.Call.Users) == 1 {
-		p.track(evCallStarted, map[string]interface{}{
-			"ParticipantID": userID,
-			"CallID":        state.Call.ID,
-			"ChannelID":     channelID,
-			"ChannelType":   channel.Type,
-		})
-
-		// new call has started
-		// If this is TestMode (DefaultEnabled=false) and sysadmin, send an ephemeral message
-		cfg := p.getConfiguration()
-		if cfg.DefaultEnabled != nil && !*cfg.DefaultEnabled &&
-			p.API.HasPermissionTo(userID, model.PermissionManageSystem) {
-			p.pluginAPI.Post.SendEphemeralPost(
-				userID,
-				&model.Post{
-					UserId:    p.botSession.UserId,
-					ChannelId: channelID,
-					Message:   "Currently calls are not enabled for non-admin users. You can change the setting through the system console",
-				},
-			)
-		}
-
-		postID, threadID, err := p.startNewCallPost(userID, channelID, state.Call.StartAt, title, threadID)
-		if err != nil {
+	} else if prevState.Call == nil {
+		if err := p.handleCallStarted(state.Call, req, channel); err != nil {
```

Comment:
The `addUserSession` -> `handleCallStarted` flow is a bit confusing to me... If I were doing it from scratch, I would want to check if there's an ongoing call, if not then start one. That seems a bit more clear?